server/src/main/java/org/elasticsearch/cluster/metadata/SingleNodeShutdownMetadata.java

```diff
@@ -0,0 +1,268 @@
+    private static final ParseField TYPE_FIELD = new ParseField("type");
+    private static final ParseField REASON_FIELD = new ParseField("reason");
+    private static final ParseField STATUS_FIELD = new ParseField("status");
+    private static final String STARTED_AT_READABLE_FIELD = "shutdown_started";
+    private static final ParseField STARTED_AT_MILLIS_FIELD = new ParseField(STARTED_AT_READABLE_FIELD + "millis");
+    private static final ParseField SHARD_MIGRATION_FIELD = new ParseField("shard_migration");
+
+    public static final ConstructingObjectParser<SingleNodeShutdownMetadata, Void> PARSER = new ConstructingObjectParser<>(
+        "node_shutdown_info",
+        a -> new SingleNodeShutdownMetadata((String) a[0], (String) a[1], (String) a[2], Status.valueOf((String) a[3]), (long) a[4],
+            (ComponentShutdownStatus) a[5])
+    );
+
+    static {
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), NODE_ID_FIELD);
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), TYPE_FIELD);
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), REASON_FIELD);
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), STATUS_FIELD);
+        PARSER.declareLong(ConstructingObjectParser.constructorArg(), STARTED_AT_MILLIS_FIELD);
+        PARSER.declareObject(
+            ConstructingObjectParser.constructorArg(),
+            (parser, context) -> ComponentShutdownStatus.parse(parser),
+            SHARD_MIGRATION_FIELD
+        );
+    }
+
+    public static SingleNodeShutdownMetadata parse(XContentParser parser) {
+        return PARSER.apply(parser, null);
+    }
+
+    private final String nodeId;
+    private final String type;
+    private final String reason;
+    private final Status status;
+    private final long startedAtDate;
+    private final ComponentShutdownStatus shardMigrationStatus;
+
+
+    public SingleNodeShutdownMetadata(
+        String nodeId, String type, String reason, Status status, long startedAtDate, ComponentShutdownStatus shardMigrationStatus) {
+        this.nodeId = Objects.requireNonNull(nodeId, "node ID must not be null");
+        this.type = Objects.requireNonNull(type, "shutdown type must not be null");
+        this.reason = Objects.requireNonNull(reason, "shutdown reason must not be null");
+        this.status = status;
+        this.startedAtDate = startedAtDate;
+        this.shardMigrationStatus = Objects.requireNonNull(shardMigrationStatus, "shard migration status must not be null");
+    }
+
+    public SingleNodeShutdownMetadata(StreamInput in) throws IOException {
+        this.nodeId = in.readString();
+        this.type = in.readString();
+        this.reason = in.readString();
+        this.status = in.readEnum(Status.class);
+        this.startedAtDate = in.readVLong();
+        this.shardMigrationStatus = new ComponentShutdownStatus(in);
```

Comment:
There seems like a disconnect here. It seems like shard migration is *one* component, but not all the components of shutting down. Rather, I think we want a `List` or `Map` of `ComponentShutdownStatus` objects, where each one has a name. Then we have one for shard migration, one for plugin signaling, one for persistent tasks, etc.

So something like

```
  "status": {
    "shard_migration": {
       "status": "IN_PROGRESS",
       "shard_migrations_remaining": 7,
       "time_started": "<user readable date>",
       "time_started_millis": 234091892
     },
     "persistent_tasks": {
       "status": "IN_PROGRESS",
       "tasks_remaining": 2,
       "error": "ICouldntStopTheTasksException[i can't do that dave]...etc stacktrack etc...",
       "time_started": "<user readable date>",
       "time_started_millis": 128391987
     },
     "plugins": {
       "status": "NOT_STARTED",
     },
```

Where each of those either a `ComponentShutdownStatus` (or maybe a subclass in the future if we want to have different specific info for each type, but that's getting ahead of this PR)

Reply:
> It seems like shard migration is one component, but not all the components of shutting down.

Right - what's there for the shard migration status was more about making sure that once the `ComponentShutdownStatus` class was added, all the serialization, diffing, etc. is still handled correctly. My plan was to add new `ComponentShutdownStatus` fields as necessary once we actually develop those components (along with exactly what diagnostic info each task needs, e.g. `shard_migrations_remaining`).

One of the reasons I didn't do so already is that I'm not sure if it would be better to use something like a `List` or `Map` there, or if the components/phases/whatever are going to be known in advance to the point where just hardcoding them would be a better option.

There's going to be some awkwardness whichever way we go, I think - if we use a container like `Map`, we'll have to deal with initializing that map, which will likely just be basically hardcoded anyway. The central code could be a bit simpler if we just use a map or similar though, as iterating through the component status objects would be trivial.

Do you have any thoughts? I'm happy to adjust this either way, just not sure which way would be best.

Reply:
We discussed this elsewhere and decided to just hardcode these, as we don't expect to need more in the future, at least to the degree where it would be worth the extra complexity to make it more flexible.